### PR TITLE
RDNSS: support more ipv6 addresses

### PR DESCRIPTION
**Pull request by WenChao1Hou**
RFC 6106 recommended that the number of RDNSS addresses thatshould be learned and maintained through the RDNSS RA option should be limited to three.  RFC 8106 removes that recommendation; thus, the number of RDNSS addresses to maintain is determined by an implementer's local policy.The number of RDNSS addresses to maintain is determined by the user's own profile

Fix:[#182](https://github.com/radvd-project/radvd/issues/182)

### CODE REVIEW

File: send.c [hunk 1]
```diff
@@ -594,6 +596,12 @@ static struct safe_buffer_list *add_ra_options_rdnss(struct safe_buffer_list *sb
 
 		memset(&rdnssinfo, 0, sizeof(rdnssinfo));
 
+		size_t const bytes = sizeof(rdnssinfo) + sizeof(struct in6_addr) * rdnss->AdvRDNSSNumber;
+		if (bytes > (256 * 8)) {
```
Comment (robbat2) [inline send.c:600]: Why did you pick 2048? 
- 1280 bytes is the min IPv6 MTU; what would happen in that case?
- can radvd safely send multiple RA packets with different RDNSS entries? (other framework code for example splits large options into their own packets)

**Review by robbat2 — APPROVED** [review]
looks mostly good, some thoughts added; will wait for CI validation.

**Comment by robbat2** [comment]
Please amend the commit to include a `Signed-Off-By` trailer per DCO.

**Comment by robbat2** [comment]
Compile failure, wonder if you missed something?
```
send.c:620:31: error: use of undeclared identifier 'rdnssbuf'
                safe_buffer_append(sbl->sb, rdnssbuf->buffer, rdnssbuf->used);
                                            ^
send.c:620:49: error: use of undeclared identifier 'rdnssbuf'
                safe_buffer_append(sbl->sb, rdnssbuf->buffer, rdnssbuf->used);
                                                              ^
send.c:621:20: error: use of undeclared identifier 'rdnssbuf'
                safe_buffer_free(rdnssbuf);
                                 ^
```

**Comment by robbat2** [comment]
I've got some tests kicking around to add as well after you have it fixed.

**Comment by stappersg** [comment]
> Please amend the commit to include a `Signed-Off-By` trailer per DCO.

@WenChao1Hou that is up-to-you. So when it is day again in your time zone, please add `Signed-Off-By`.

**Comment by WenChao1Hou** [comment]
> Compile failure, wonder if you missed something?
> 
> ```
> send.c:620:31: error: use of undeclared identifier 'rdnssbuf'
>                 safe_buffer_append(sbl->sb, rdnssbuf->buffer, rdnssbuf->used);
>                                             ^
> send.c:620:49: error: use of undeclared identifier 'rdnssbuf'
>                 safe_buffer_append(sbl->sb, rdnssbuf->buffer, rdnssbuf->used);
>                                                               ^
> send.c:621:20: error: use of undeclared identifier 'rdnssbuf'
>                 safe_buffer_free(rdnssbuf);
>                                  ^
> ```



> Compile failure, wonder if you missed something?
> 
> ```
> send.c:620:31: error: use of undeclared identifier 'rdnssbuf'
>                 safe_buffer_append(sbl->sb, rdnssbuf->buffer, rdnssbuf->used);
>                                             ^
> send.c:620:49: error: use of undeclared identifier 'rdnssbuf'
>                 safe_buffer_append(sbl->sb, rdnssbuf->buffer, rdnssbuf->used);
>                                                               ^
> send.c:621:20: error: use of undeclared identifier 'rdnssbuf'
>                 safe_buffer_free(rdnssbuf);
>                                  ^
> ```

it was an oversight on my part，The variable name was not updated in time when submitted, I will update MR as soon as possible

File: send.c (hunk 1 above)
Comment (WenChao1Hou) [reply send.c:600]: in my opinion
1、The IPv6 MTU value may change the processing of the kernel protocol stack.
2、RDNSS 2048 is consistent with DNSSL packet maximum in radvd

Do you have any good suggestions for me?

**Comment by the-j0k3r** [comment]
We had merged this patch downstream at this code state https://svn.dd-wrt.com/changeset/50755, I did not test it myself but clearly it caused some instability and was reverted

One such report https://forum.dd-wrt.com/phpBB2/viewtopic.php?p=1275031#1275031

>Same thing happened to me. If you clean those two ipv6 dns lines (leave them empty) radvd deamon will not crash. Your/my /tmp/radvd.conf look/ed like:

>root@tp:/tmp# cat radvd.conf
>interface br0
>{
>IgnoreIfMissing on;
>AdvSendAdvert on;
>MinRtrAdvInterval 3;
>MaxRtrAdvInterval 10;
>AdvHomeAgentFlag off;
>AdvManagedFlag off;
>AdvOtherConfigFlag on;
>AdvLinkMTU 1452;
>prefix ::/64
>{
>AdvOnLink on;
>AdvAutonomous on;
>AdvValidLifetime 30;
>AdvPreferredLifetime 20;
>};
>RDNSS 2606:4700:4700::1111 2606:4700:4700::1001{};
>};

>On 50786 once you leave those boxes empty you'll end up whith the same conf sans that RDNSS line and radvd won´t die.

>Syslog entries of radvd deamon dying:

>Nov 3 07:07:12 tp daemon.info radvd[9224]: version 2.19 started
>Nov 3 07:07:12 tp daemon.err radvd[9226]: Exiting, privsep_read_loop had readn return 0 bytes
>Nov 3 07:07:12 tp daemon.err radvd[9226]: Exiting, privsep_read_loop is complete.

>Previous firm with that line working: 50485.

@wcbonner @BrainSlayer

**Comment by reubenhwk** [comment]
Just catching up here....why are more than 3 RDNS servers needed?  I saw that more than three doesn't work, but I did NOT see any explanation as to why more than three are needed.

**Comment by wcbonner** [comment]
> Just catching up here....why are more than 3 RDNS servers needed?  I saw that more than three doesn't work, but I did NOT see any explanation as to why more than three are needed.

There isn't really a huge need for more than three, but it would be nice if the program didn't crash or error out if more are listed in the config file. 

The problem I ran into was in a downstream project when I wanted to specify two dns servers and the program was already adding two servers it had received from another network broadcast or dhcp interaction. 

radvd was silently exiting in the background and my local network wasn't getting advertisements.

**Comment by reubenhwk** [comment]
> 

Ack.  It's been a long time since I've work on this code... Nonetheless I have concerns that not having a hard limit would eventually lead to MTU issues.  At the same time I'm not sure that exiting the program is also the right thing to do.  Not sure.  Maybe best effort is good enough, maybe not.

**Comment by Neustradamus** [comment]
Dear all,

All people are welcome to comment the current situation, we are in December 2024, a perfect patch is needed.

It is really important to respect the RFC8106: IPv6 Router Advertisement Options for DNS Configuration:
- https://datatracker.ietf.org/doc/html/rfc8106

@agowa, @AdSchellevis, @fichtner, @marjohn56, @bmeirellesRJ, @mhanor, @stu-gott, @wcbonner, @the-j0k3r, @WenChao1Hou, @BrainSlayer, @Mile-Lile, @bnutzer, @stu-gott, @MarioVerbelen, @ruben-herold, @ipilcher, @pyk1998, @romanrm, @corvus1, @ihipop

If you know some IPv6 specialists, please come here, there are several tickets, thanks in advance.

**Comment by Neustradamus** [comment]
@robbat2, @reubenhwk, @stappersg, @RICCIARDI-Adrien, @y-kzm, @fichtner, @melroy89, @xdbob, @audreylace, @XWwalker, @initramfs, @thesamesam, @jetomit, @mpontillo, @kepstin, @oskar456, @zajdee, @yas-nyan, @gaoxingwang, @juhamatk, @manonfgoo, @andrew-sayers, @jpds, @steglicd, @samboyles1, @nomis, @haegar, @ezeeyahoo, @Testsr, @landgraf, @hxx-fetch, @prometheanfire, @tlsalmin, @ffontaine, @BrainSlayer, @Chocobo1, @sshambar, @bluca, @dhallas, @florianl, @fooker, @trofi, @l29ah, @thediveo, @smlng, @candrews, @alexaring, @steglicd, @pavlix, @Erkan-Yilmaz, @abrodkin, @crrodriguez, @neocturne, @danrl, @vapier, @GheRivero, @segoon, @moonlinux, @sean797, @ferdinandoterada, @fgont, @nachtgeist, all others: What do you think?

All people are welcome to comment, help...

If you know other IPv6 experts, specialists, it will be nice too.

This PR needs to be followed by this @robbat2 PR:
- https://github.com/radvd-project/radvd/pull/195

Linked to the next 2.20 release build:
- https://github.com/radvd-project/radvd/issues/185

Some other tickets can be solved too:
- https://github.com/radvd-project/radvd/issues/185#issuecomment-2494498563

Thanks in advance, we need to have a good 2.20 since 2.19 (2020-09-23).

**Comment by Neustradamus** [comment]
+ @crrodriguez, @neocturne, @danrl, @vapier, @GheRivero, @segoon, @moonlinux, @sean797, @ferdinandoterada, @fgont, @nachtgeist, all others: What do you think?

**Comment by danrl** [comment]
As a compromise that should work for most reasonable cases I can imagine right now radvd could behave the following way:

- Exit at startup time when a config asks for more RDNSS than would fit inside a RA limited by minimum MTU 1280 bytes. This should be at config time for static setups. Strictly speaking the RFC would require additional RAs to be send to fit an excessive number of RDNSS but all RA daemons I have seen put limits in, even my own ratools suite which was written for academic purpose of following the RFC absurdly strictly.
- Continue operating in other cases of excess, but log an error periodically, maybe in RA announcement intervals.

One could argue that randomizing the list of excessive RDNSS each time a RA is sent and then picking as many from the top as fit within 1280 bytes would eventually allow all RDNSS to he announced at least once. However, this makes RADVD unpredictable in behavior which is bad. Predictable is important for reliable testing and debugging. It is more important than following the letter of the RFC. Well tested, reasonable behavior is better than covering all use cases.

My suggestion for moving forward: Limit RDNSS to a reasonable number. 2048 is too many. We can barely fit 70 in there even if we could use the entire packet minus header for 16 bytes addresses. Now how many RDNSS can we fit in for a reasonable scenario? Maybe one where RDNSS is the only option and we don’t even have a Prefix option. That is the most somewhat reasonable scenario I can think of. The limit should be at max this number. The math shall be provided in the commit message for future archeology. 

HTH

**Comment by melroy89** [comment]
I would expect a test in: https://github.com/radvd-project/radvd/tree/master/test

**Comment by robbat2** [comment]
Replaced by https://github.com/radvd-project/radvd/pull/245

**Pull request closed without merging**